### PR TITLE
Fix nodesDir scan when node package has js/html in sub dir to package.json

### LINKED ISSUE #3861: nodesDir crashes Node-RED 3.0
### Current Behavior

I’m developing custom nodes in a directory specified with the **nodesDir** property in the settings.js configuration file. This set-up used to work up to Node-RED 2.2.2. However, I’ve just upgraded to Node-RED 3.0.2, and now the server crashes immediately on start-up, using a settings.js file that used to work before the upgrade.

### Expected Behavior

When the **nodesDir** property is used in the settings.js to specify a custom nodes directory, the server should not crash. If a problem is encountered, then the nature of this problem shall be described in the log output.

### Steps To Reproduce

If I use the following [settings.js](https://www.celtrio.com/varenne/node-red/ok-settings.js), Node-RED 3.0.2 starts normally. Please note that some values have been redacted in this file for confidentiality reasons. This first settings.js file works with Node-RED 2.2.2

However, if I use this second [settings.js](https://www.celtrio.com/varenne/node-red/ko-settings.js) file, then the Node-RED 3.0.2 server crashes on start-up with the following output:

```
Welcome to Node-RED
===================
1 Sep 20:18:04 - [info] Node-RED version: v3.0.2
1 Sep 20:18:04 - [info] Node.js  version: v16.17.0
1 Sep 20:18:04 - [info] Linux 5.10.0-9-amd64 x64 LE
1 Sep 20:18:04 - [info] Loading palette nodes
1 Sep 20:18:04 - [debug] Module: varenne-nodes 1.0beta1 /home/node-red/varenne-nodes
1 Sep 20:18:04 - [error] Failed to start server:
1 Sep 20:18:04 - [error] TypeError: Cannot read properties of undefined (reading 'forEach')
    at Object.addModule (/usr/lib/node_modules/node-red/node_modules/@node-red/registry/lib/registry.js:195:27)
    at /usr/lib/node_modules/node-red/node_modules/@node-red/registry/lib/loader.js:153:34
```

The only difference between these two settings.js files is the following line, which is present when the server crashes.

`nodesDir: '/home/node-red/varenne-nodes'`

### Example flow

With **nodesDir** set, Node-RED 3.0.2 crashes on start-up, even with an empty flow.

### Environment

- Node-RED version: 3.0.2
- Node.js version: v16.17.0
- npm version: 8.19.0
- Platform/OS: Linux 5.10.0-9-amd64 #1 SMP Debian 5.10.70-1 (2021-09-30) x86_64 GNU/Linux
- Browser: Not browser related

### PR DESCRIPTION
**Pull request by Steve-Mcl**
fixes #3861 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

First, please read through the issue detailed in [this comment](https://github.com/node-red/node-red/issues/3861#issuecomment-1236124417)

## Proposed changes

* Normalise path before splitting (`loader.js`)
* Skip calling statSync on empty string  (`loader.js`)
* Shortcut scanning of a directories files/folders if the directory contains a package.json (`localfilesystem.js`)
* Prevent crashing node-red when `set.types` is not set  (`registry.js`)
* Print out which node caused the error 
    * BEFORE: 
    ```
    TypeError: Cannot read properties of undefined (reading 'forEach')
    at Object.addModule (/usr/lib/node_modules/node-red/node_modules/@node-red/registry/lib/registry.js:195:27)
    at /usr/lib/node_modules/node-red/node_modules/@node-red/registry/lib/loader.js:153:34
    ```
    * AFTER:    (Updated 2022/09/29, after commit e6b3793)
    ```
    29 Sep 21:25:17 - [warn] [node-red/lower-case] Set does not have any types. name: 'lower-case', module: 'node-red', file: 'C:\Users\user\repos\_node-red-issue-3861\lower-case.js'
    packages/node_modules/@node-red/util/lib/log.js:94
    29 Sep 21:25:31 - [warn] [node-red-node-google/google-api-config] 'google-api-config' already registered by module node-red-contrib-google-calendar
    packages/node_modules/@node-red/util/lib/log.js:94
    29 Sep 21:25:31 - [warn] [node-red-node-rbe/rbe] 'rbe' already registered by module node-red
    packages/node_modules/@node-red/util/lib/log.js:94
    29 Sep 21:25:31 - [warn] [@flowforge/nr-project-nodes/project-link] Error: Project Link nodes cannot be loaded outside of flowforge EE environment
    packages/node_modules/@node-red/util/lib/log.js:94
    29 Sep 21:25:57 - [warn] [lower-case/lower-case] Set does not have any types. name: 'lower-case', module: 'lower-case', file: 'C:\Users\user\repos\_node-red-issue-3861\lower-case\lower-case\lower-case.js'
    packages/node_modules/@node-red/util/lib/log.js:94
    29 Sep 21:26:06 - [warn] [@lowercase/lower-case2/lower-case] Set does not have any types. name: 'lower-case', module: '@lowercase/lower-case2', file: 'C:\Users\user\repos\_node-red-issue-3861\@lower-case\lower-case2\lower-case.js'
    ```

* Add more tests that load nodes with js/html files in a sub directory (`registry.js`)

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [ ] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [x] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

File: packages/node_modules/@node-red/registry/lib/registry.js [hunk 1]
```diff
@@ -185,10 +185,19 @@ function loadNodeConfigs() {
...
         if (module.nodes.hasOwnProperty(setName)) {
             var set = module.nodes[setName];
+            if (!set.types) {
+                const err = new Error("Node has no types")
+                err.code = "has_no_types"
+                err.details = {
+                    setName: set.name,
+                    set: { ...set }
+                }
+                console.warn(err)
```
Comment (knolleary) [inline packages/node_modules/@node-red/registry/lib/registry.js:198]:
```suggestion
                console.warn(err)
```

**Review by knolleary — COMMENTED** [review]
Your before/after shows the log output in the before, but a screenshot of the error object inside VSCode for the after.
End-users won't be running in VSCode - what actually gets logged? Does it include the extra information about what module caused it?

File: packages/node_modules/@node-red/registry/lib/registry.js (hunk 1 above)
Comment (Steve-Mcl) [reply packages/node_modules/@node-red/registry/lib/registry.js:198]: I'm not certain at this moment in time (I forget if `warn` actually outputs the extra detail). I Will need to run it again.

Comment (Steve-Mcl) [reply packages/node_modules/@node-red/registry/lib/registry.js]: ok, so I think i have a better solution for logging this.

1. add a templated i18n message
2. dont log the error at the time of finding the `set` has no `types` - simply set the `err.details` property
3. in `runtime/lib/index:162` (where `"type_already_registered"` is also logged), log an i18n ready warning message from the `err.details`

This results in the following logs...

```
29 Sep 21:25:17 - [warn] [node-red/lower-case] Set does not have any types. name: 'lower-case', module: 'node-red', file: 'C:\Users\user\repos\_node-red-issue-3861\lower-case.js'
packages/node_modules/@node-red/util/lib/log.js:94
29 Sep 21:25:31 - [warn] [node-red-node-google/google-api-config] 'google-api-config' already registered by module node-red-contrib-google-calendar
packages/node_modules/@node-red/util/lib/log.js:94
29 Sep 21:25:31 - [warn] [node-red-node-rbe/rbe] 'rbe' already registered by module node-red
packages/node_modules/@node-red/util/lib/log.js:94
29 Sep 21:25:31 - [warn] [@flowforge/nr-project-nodes/project-link] Error: Project Link nodes cannot be loaded outside of flowforge EE environment
packages/node_modules/@node-red/util/lib/log.js:94
29 Sep 21:25:57 - [warn] [lower-case/lower-case] Set does not have any types. name: 'lower-case', module: 'lower-case', file: 'C:\Users\user\repos\_node-red-issue-3861\lower-case\lower-case\lower-case.js'
packages/node_modules/@node-red/util/lib/log.js:94
29 Sep 21:26:06 - [warn] [@lowercase/lower-case2/lower-case] Set does not have any types. name: 'lower-case', module: '@lowercase/lower-case2', file: 'C:\Users\user\repos\_node-red-issue-3861\@lower-case\lower-case2\lower-case.js'
```

Resolved in [e6b3793](https://github.com/node-red/node-red/pull/3867/commits/e6b379358aa1278d261bc7bee2a1be09f9fc41f0)

File: packages/node_modules/@node-red/registry/lib/registry.js [hunk 2]
```diff
@@ -185,10 +185,20 @@ function loadNodeConfigs() {
...
     moduleConfigs[module.name] = module;
-    // console.log("registry.js.addModule",module.name,"user?",module.user,"usedBy",module.usedBy,"dependencies",module.dependencies)
-    for (var setName in module.nodes) {
+    for (const setName in module.nodes) {
         if (module.nodes.hasOwnProperty(setName)) {
-            var set = module.nodes[setName];
+            const set = module.nodes[setName];
+            if(setName === 'lower-case') {
+                delete set.types
+            }
```
Comment (knolleary) [inline packages/node_modules/@node-red/registry/lib/registry.js:193]: Is this test/debugging code?

Comment (Steve-Mcl) [inline packages/node_modules/@node-red/registry/lib/registry.js:193]:
```suggestion
```

Comment (Steve-Mcl) [reply packages/node_modules/@node-red/registry/lib/registry.js]: resolved in [6d1a12a](https://github.com/node-red/node-red/pull/3867/commits/6d1a12af4ba3c26a221962eb7aa56dbe129375d7)

**Pull request merged**